content: Mention containerd 1.7 supports user namespaces (#40264)
kubernetes/website: content/en/docs/concepts/workloads/pods/user-namespaces.md

<!-- body -->
## {{% heading "prerequisites" %}}

{{% thirdparty-content %}}

This is a Linux only feature. In addition, support is needed in the 
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to use this feature with Kubernetes stateless pods:

* CRI-O: version 1.25 (and later) supports user namespaces for containers.

* containerd: version 1.7 supports user namespaces for containers, compatible

>   with Kubernetes {{< skew currentVersion >}}. If you are running a different

Author: This will break if we apply this to 1.25. This will point to kubernetes 1.24, IIUC, which simply doesn't support user namespaces

Reviewer: Yep, I don't recommend backporting _this_ line to the v1.25 docs

Author: IMHO it will be simpler if we just use the versions, as they are not relative to any k8s version. In fact, here just expands to k8s 1.26, but this is true with k8s 1.25 too.

Reviewer 2: I'd prefer calling out the specific version in this context.
We say "the WWII ended in 1945" instead of "in {{ current year }}, the WWII has ended." with a constant refreshing of the `{{ current year }}`.

Reviewer: Historical events aren't quite the right analogy. It's about a commutative relationship between (releases of) two things. The containerd docs might want to express the reverse relationship for example.

The text I've proposed for v1.27 also uses a skew shortcode (for a different reason).

I guess if we're sure that v1.7 of containerd is compatible specifically with v1.26 of Kubernetes and no other minor version of this project, we _should_ say exactly that (and keep the “If you are running a different version of Kubernetes, check the documentation for that Kubernetes release.”)

However, this is an unusual case because of when we're making the change, ie after the code freeze for v1.27

Author: @sftim what change is after code freeze? This doc PR?

First, this doc isn't for 1.27. This is for 1.25 and 1.26.

Secondly, that statement is true for 1.25 and 1.26, I don't understand why you insist in mentioning only one release.

And yes, we re-worked the implementation in 1.27 and what we need other requirements from the runtime. That is why it is not recommended to use 1.7 for userns with k8s >= 1.27.

But to return to more practical matter, how do you want me to phrase it here? So we can merge this PR

Reviewer: I commented: https://github.com/kubernetes/website/pull/40264/files#r1149286339

Author: @sftim that link is not working (github is having hiccups, maybe is due to that). Can you please explain here again? Or a suggestion is even better ;)

<!-- body -->
## {{% heading "prerequisites" %}}

{{% thirdparty-content %}}

<!-- When merging this with the dev-1.27 branch conflicts will arise. The text
as it is in the dev-1.27 branch should be used. -->
This is a Linux only feature. In addition, support is needed in the 
{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to use this feature with Kubernetes stateless pods:

* CRI-O: version 1.25 (and later) supports user namespaces for containers.

* containerd: version 1.7 supports user namespaces for containers, compatible
  with Kubernetes {{< skew currentVersion >}}. If you are running a different

>   version of Kubernetes, check the documentation for that Kubernetes release.

Author: Please feel free to reopen if some of those changes don't seem right for you.

> {{% thirdparty-content single="true" %}}
> <!-- if adding another runtime in the future, omit the single setting -->

Reviewer: ```suggestion
{{% thirdparty-content %}}
```

> This is a Linux only feature. In addition, support is **needed** in the 

Reviewer: That change doesn't line up with the PR title, and I also disagree that it's required. Separately, I think it could be OK to change the wording to help people who don't read documentation understand the requirement. But not in this commit.

Author: removed

{{< glossary_tooltip text="container runtime" term_id="container-runtime" >}}
to use this feature with Kubernetes stateless pods:

> * CRI-O: supports user namespaces with all Kubernetes versions >= v1.25

Reviewer: ```suggestion
* CRI-O: version 1.25 (and later) of CRI-O supports user namespaces for containers.
```

Using CRI-O v1.24 and Kubernetes v1.26? It won't work (AIUI). It's the CRI-O version that matters here.

Author: I can change it to that, but CRIO is released in lockstep with k8s. You can't use CRIO 1.24 with k8s 1.26 anyways, AFAIK.

> * containerd: v1.7 supports user namespace with Kubernetes v1.25 and v1.26.

Reviewer 3: ```suggestion
* containerd: v1.7 supports user namespace with Kubernetes v1.25 and above.
```

1.27 will keep working

> * containerd: v1.7 supports user namespace with Kubernetes v1.25 and v1.26.

Reviewer: ```suggestion
* containerd: version 1.7 of CRI-O supports user namespaces for containers,
  compatible with Kubernetes {{< skew currentVersion >}}. If you are running a different
  version of Kubernetes, check the documentation for that Kubernetes release.
```

This approach works, regardless of what happens with future containerd and Kubernetes releases. I strongly recommend using it here.

Author: Added, changing the typo: s/of CRI-O//

Containerd plans to add support for user namespaces with Kubernetes >= 1.27 in

> version 1.8. For more information, see the [containerd issue][containerd-issue].

Reviewer: We [avoid statements about the future](url). _Expecially_ about third party projects.

We should drop this text.

Author: ok, removed

Support for this in [cri-dockerd is not planned][CRI-dockerd-issue] yet.

You should not use user namespaces support in Kubernetes >= 1.27 with
containerd 1.7. The Kubernetes implementation was reworked and has requirements

> that containerd 1.7 doesn't support.

Reviewer: That change needs to land on the branch _dev-1.27_ and not here.

Author: ok, removed

> [containerd-issue]: https://github.com/containerd/containerd/issues/8286

Reviewer: ```suggestion
```
No need.
